src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelperFactory.cs

```diff
@@ -9,9 +12,40 @@ namespace Microsoft.AspNetCore.Mvc.Routing
     public class UrlHelperFactory : IUrlHelperFactory
     {
         /// <inheritdoc />
+        /// Perf: Added a pool of StringBuilder that will be used by URL Helpers when generating URLs for requests
+        private ObjectPool<StringBuilder> _stringBuilderPool;
+
+        public ObjectPool<StringBuilder> StringBuilderPool
+        {
+            get
+            {
+                if (_stringBuilderPool == null)
+                {
+                    _stringBuilderPool = new DefaultObjectPoolProvider().Create<StringBuilder>();
+                }
+
+                return _stringBuilderPool;
+            }
+        }
         public IUrlHelper GetUrlHelper(ActionContext context)
         {
-            return new UrlHelper(context);
+            var httpContext = context.HttpContext;
+            IUrlHelper urlHelper = null;
+            if (httpContext != null && httpContext.Features != null)
+            {
+                urlHelper = httpContext.Features[typeof(IUrlHelper)] as IUrlHelper;
```

Comment:
This should go in `HttpContext.Items` - features is something different entirely